lib/Listener/BeforeNodeReadListener.php

```diff
@@ -0,0 +1,138 @@
+<?php
+
+declare(strict_types=1);
+
+/**
+ * SPDX-FileCopyrightText: 2025 Nextcloud GmbH and Nextcloud contributors
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+namespace OCA\Files_DownloadLimit\Listener;
+
+use OCA\Files_DownloadLimit\Db\LimitMapper;
+use OCP\AppFramework\Db\DoesNotExistException;
+use OCP\EventDispatcher\Event;
+use OCP\EventDispatcher\IEventListener;
+use OCP\Files\Events\BeforeZipCreatedEvent;
+use OCP\Files\Events\Node\BeforeNodeReadEvent;
+use OCP\Files\File;
+use OCP\Files\Folder;
+use OCP\Files\NotFoundException;
+use OCP\Files\Storage\ISharedStorage;
+use OCP\Share\IManager;
+use OCP\Share\IShare;
+use Psr\Log\LoggerInterface;
+
+/**
+ * @template-implements IEventListener<BeforeNodeReadEvent|BeforeZipCreatedEvent|Event>
+ */
+class BeforeNodeReadListener implements IEventListener {
+	public function __construct(
+		private IManager $manager,
+		private LimitMapper $mapper,
+		private LoggerInterface $logger,
+	) {
+	}
+
+	public function handle(Event $event): void {
+		if ($event instanceof BeforeZipCreatedEvent) {
+			$this->handleBeforeZipCreatedEvent($event);
+		} elseif ($event instanceof BeforeNodeReadEvent) {
+			$this->handleBeforeNodeReadEvent($event);
+		}
+	}
+
+	public function handleBeforeZipCreatedEvent(BeforeZipCreatedEvent $event): void {
+		$files = $event->getFiles();
+		if (count($files) !== 0) {
+			/* No need to do anything, count will be triggered for each file in the zip by the BeforeNodeReadEvent */
+			return;
+		}
+
+		$node = $event->getFolder();
+		if (!($node instanceof Folder)) {
+			return;
+		}
+
+		try {
+			$storage = $node->getStorage();
+		} catch (NotFoundException) {
+			return;
+		}
+
+		if (!$storage->instanceOfStorage(ISharedStorage::class)) {
+			return;
+		}
+
+		/** @var ISharedStorage $storage */
+		$share = $storage->getShare();
+
+		if ($share->getShareType() !== IShare::TYPE_LINK) {
+			return;
+		}
+
+		$this->singleFileDownloaded($share);
+	}
+
+	public function handleBeforeNodeReadEvent(BeforeNodeReadEvent $event): void {
+		$node = $event->getNode();
+		if (!($node instanceof File)) {
+			return;
+		}
+
+		try {
+			$storage = $node->getStorage();
+		} catch (NotFoundException) {
+			return;
+		}
+
+		if (!$storage->instanceOfStorage(ISharedStorage::class)) {
+			return;
+		}
+
+		/** @var ISharedStorage $storage */
+		$share = $storage->getShare();
+
+		if ($share->getShareType() !== IShare::TYPE_LINK) {
+			return;
+		}
+
```

Comment:
The check for range requests and cache is missing, that means when playing a video it will count as a download for each request.
I think we should have here the same thing as activity to only count one download for a video play.